substrate-archive/src/actors/workers/extrinsics_decoder.rs

```diff
@@ -0,0 +1,174 @@
+// Copyright 2017-2021 Parity Technologies (UK) Ltd.
+// This file is part of substrate-archive.
+
+// substrate-archive is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+// substrate-archive is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with substrate-archive.  If not, see <http://www.gnu.org/licenses/>.
+
+use arc_swap::ArcSwap;
+use async_std::task;
+use itertools::Itertools;
+use sqlx::PgPool;
+use std::{collections::HashMap, sync::Arc};
+use xtra::prelude::*;
+
+use desub::Decoder;
+
+use crate::{
+	actors::{
+		workers::database::{DatabaseActor, GetState},
+		SystemConfig,
+	},
+	database::{models::ExtrinsicsModel, queries},
+	error::{ArchiveError, Result},
+	types::BatchExtrinsics,
+};
+
+/// Actor which crawls missing encoded extrinsics and
+/// sends decoded JSON to the database.
+/// Crawls missing extrinsics upon receiving an `Index` message.
+pub struct ExtrinsicsDecoder {
+	/// Pool of Postgres Connections.
+	pool: PgPool,
+	/// Address of the database actor.
+	addr: Address<DatabaseActor>,
+	/// Max amount of extrinsics to load at any one time.i.
+	max_block_load: u32,
+	/// Desub Legacy + current decoder.
+	decoder: Arc<Decoder>,
```

Comment:
At the moment, it's not clear to me why we need to surround this (and below) in Arc? As far as I can see (very much offhand and I'm almost certain I'm missing something at the mo) this can't be cloned, and is probably a single long running actor that doesn't need synchronisation things?

Reply:
Aah I see; it's for spawning a copy of this onto a task below; got it!